### ISSUE TITLE
Throwing in a Results listener callback crashes the app

### ISSUE THREAD

**Issue by kraenhansen** [realm/realm-js#2654]
## Goals

Realm should never crash a React Native app.

I suspect that we need to tell the JSCore engine that an unhandled exception happened during the execution of the callback.

## Expected Results

When an error is thrown from a callback added as a listener on a `Results` object, I expect an uncaught exception to be handled (resulting in a black screen with a red top-bar while in development mode).

## Actual Results

When an error is thrown from a callback added as a listener on a `Results` object, the app crashes!

## Steps to Reproduce & Code Sample

```javascript
// App.js

import React, { Component } from 'react';
import { View, Text } from 'react-native';
import Realm from 'realm'

export default class App extends Component {
  componentDidMount() {
    const realm = new Realm({
        schema: [{
            name: "Singleton",
            properties: { value: "string" },
        }],
    });

    realm.objects("Singleton").addListener(() => {
        throw new Error("Change happened!");
    });
  }

  render() {
    return <View><Text>App loaded ...</Text></View>;
  }
}
```

## Version of Realm and Tooling

- Realm JS SDK Version: 2.6.0
- Node or React Native: React Native (confirmed on iOS)
- Client OS & Version: N/A
- Which debugger for React Native: None

**Comment by kneth** [comment]
It might be out of the scope of Realm as a library to make the decision to swallow exceptions. Other options exist within the React Native framework:

* `ErrorUtils.setGlobalHandler()` - https://stackoverflow.com/questions/30918349/how-to-catch-uncaught-exception-globally-in-react-native
* `NativeModules.ExceptionsManager` - https://stackoverflow.com/questions/44774247/how-to-disable-red-screen-in-debug-mode

I don't consider the current behaviour as bug.

**Comment by kraenhansen** [comment]
> Realm should never crash a React Native app.

It might not be straight forward to understand or fix, but let's not kid ourselves: Of cause it's a bug when a library breaks invariants of the environment, namely "JS code might throw unhandled exceptions, but it will never crash the process running the JS code". It is super unexpected behavior on our part.

**Comment by kneth** [comment]
I agree that the correct behaviour is that a JS exception is thrown: a raw C++ exception is definitely not adding any value for the app developer.

**Comment by kneth** [comment]
After 6 months I still don't consider it a bug. Throwing an exception inside a callback is not best practise. Moreover, callbacks should have a `try`/`catch` and handle errors gracefully. We should document it clearly in our API documentation that we except users to handle errors inside the listener.

**Comment by steffenagger** [comment]
@kneth is the error actually thrown in JS-context?
If the error doesn't currently show up as a "red error screen" in React Native development mode (that's how I read the initial issue), it sounds like the app crashed at a level where developers can't `catch` the error.

**Comment by kneth** [comment]
We will document the behaviour until we have a clear idea of how to solve it.
